Upgrade to Ensuro V3

LucasCambon opened the pull request:
- Upgrade Solidity version to 0.8.30
- Upgrade EVM version to praga (same as OZ)
- Upgrade @ensuro/utils to latest
- Upgrade @ensuro/swaplibrary to latest
- Upgrade OZ to 5.5 & remove dead code contracts
- Move strategies to contracts/strategies
- Upgrade to Hardhat 3

LucasCambon committed:
Ensuro V3: Upgrade Solidity version, upgrade EVM version (praga), update @ensuro utils & swaplibrary, upgrade OZ to 5.5 & remove dead code contracts

LucasCambon committed:
Fix: fixed tests for the latests changes & removed contracts

LucasCambon committed:
Moved strategies to contracts/strategies folder & fixed tests using INFURA RPC

gnarvaja commented on contracts/mock/SingleStrategyERC4626.sol:
```diff
@@ -75,12 +77,13 @@ contract SingleStrategyERC4626 is PermissionedERC4626, IExposeStorage {
     string memory name_,
     string memory symbol_,
-    address admin_,
+    address,
```
Cambialo para que ya no reciba este parámetro. Sería simplemente cambiar los tests, este contrato no se usa más que en este repo.

gnarvaja replied:
Entiendo que ahora este contrato queda sin ningún permiso, todos los métodos abiertos. Tiene sentido para un contrato que se usa para tests.

gnarvaja commented on hardhat.config.js:
```diff
@@ -31,6 +38,7 @@ module.exports = {
   dependencyCompiler: {
     paths: [
       "@ensuro/utils/contracts/TestCurrency.sol",
+      "@ensuro/utils/contracts/TestCurrencyAC.sol",
```
Creo que sería mejor usar el TestCurrency a secas, sin access control. No me parece que necesitemos para los tests que tenga access control y los hace más complicados. 

Lo mismo en el repo de CashFlowLender.

gnarvaja commented on test/test-compound-v3-vault.js:
```diff
@@ -157,179 +157,6 @@ const SwapStableAaveV3InvestStrategyMethods = {
-        },
-      });
-      const strategy = await CompoundV3InvestStrategy.deploy(ADDRESSES.cUSDCv3, ADDRESSES.REWARDS);
-      const MultiStrategyERC4626 = await ethers.getContractFactory("MultiStrategyERC4626");
```
Esta variante, en lugar de borrarla, habría que hacer que sea usando el AccessManagedMSV. Salvo que eso ya esté probado más abajo en otra variante.

gnarvaja replied:
Ah, esta creo que ya está cubierta en CompoundV3Strategy+AccessManaged

gnarvaja commented on test/test-compound-v3-vault.js:
```diff
@@ -157,179 +157,6 @@ const SwapStableAaveV3InvestStrategyMethods = {
-    name: "AAVEV3Strategy",
-    cToken: ADDRESSES.aUSDCv3,
-    supplyToken: ADDRESSES.USDC,
-    fixture: async () => {
```
Esta variante, en lugar de borrarla, habría que hacer que sea usando el AccessManagedMSV. Salvo que eso ya esté probado más abajo en otra variante.

gnarvaja replied:
Esta me parece que no. Habría que hacer una `AAVEV3Strategy+AccessManaged`

gnarvaja commented on test/test-compound-v3-vault.js:
```diff
@@ -449,75 +276,6 @@ const variants = [
-      const specificRole = await vault.getForwardToStrategyRole(0, operation);
-      await grantRole(hre, vault.connect(admin), specificRole, user);
-    },
-  },
```
Idem, no borrar las variantes (ni los tests que testean cosas específicas de estas), adaptarlas para que trabajen con AccessManagedMSV.

gnarvaja reviewed (commented):
Revisé algunas cosas, después sigo con el resto.

LucasCambon committed:
Test improvements upgrading coverage %

LucasCambon committed:
Fix reviewed comments:
- Improve coverage % & branches %
- Add deleted variants tests from AAVE & SwapStable strategies
- Delete admin from SingleStrategyERC4626.sol

LucasCambon committed:
Deleted TestCurrencyAC

LucasCambon committed:
Fix broken test by deleted TestCurrencyAC

LucasCambon committed:
refavor: using access-manager-proxy as dependency

gnarvaja committed:
Deduplicate code + simplify

Avoids duplication of code between the variants.

Removes complexitity that came when multiple heterogenous variants were
active, now that all are AccessManaged, we don't need to mention it.

LucasCambon committed:
Deduplicate code + grantOperationAccess in utils
